build/common.sh

```diff
@@ -566,70 +626,134 @@ function kube::build::run_build_command() {
+    -- /rsyncd.sh >/dev/null
 
-    local -ra docker_cmd=(
-      "${DOCKER[@]}" run "${docker_run_opts[@]}" "${KUBE_BUILD_IMAGE}"
-    )
+  local mapped_port
+  if ! mapped_port=$("${DOCKER[@]}" port "${KUBE_RSYNC_CONTAINER_NAME}" 8730 2> /dev/null | cut -d: -f 2) ; then
+    kube:log:error "Could not get effective rsync port"
+    return 1
+  fi
 
-    kube::log::status "Syncing back _output/dockerized/bin directory from remote Docker"
-    rm -rf "${LOCAL_OUTPUT_BINPATH}"
-    mkdir -p "${LOCAL_OUTPUT_BINPATH}"
-    rm -f "${THIS_PLATFORM_BIN}"
-    ln -s "${LOCAL_OUTPUT_BINPATH}" "${THIS_PLATFORM_BIN}"
-
-    kube::build::destroy_container "${KUBE_BUILD_CONTAINER_NAME}"
-    "${docker_cmd[@]}" bash -c "cp -r ${REMOTE_OUTPUT_BINPATH} /tmp/bin;touch /tmp/finished;rm /tmp/bin/test_for_remote;/bin/sleep 600" > /dev/null 2>&1
-
-    # Wait until binaries have finished coppying
-    count=0
-    while true;do
-      if "${DOCKER[@]}" cp "${KUBE_BUILD_CONTAINER_NAME}:/tmp/finished" "${LOCAL_OUTPUT_BINPATH}" > /dev/null 2>&1;then
-        "${DOCKER[@]}" cp "${KUBE_BUILD_CONTAINER_NAME}:/tmp/bin" "${LOCAL_OUTPUT_SUBPATH}"
-        break;
-      fi
+  local container_ip
+  container_ip=$("${DOCKER[@]}" inspect --format '{{ .NetworkSettings.IPAddress }}' "${KUBE_RSYNC_CONTAINER_NAME}")
+
+  if kube::build::probe_address 127.0.0.1 ${mapped_port}; then
+    KUBE_RSYNC_ADDR="127.0.0.1:${mapped_port}"
+    sleep 0.5
+    return 0
+  elif kube::build::probe_address "${container_ip}" 8730; then
+    KUBE_RSYNC_ADDR="${container_ip}:8730"
+    sleep 0.5
+    return 0
+  fi
 
-      let count=count+1
-      if [[ $count -eq 60 ]]; then
-        # break after 5m
-        kube::log::error "Timed out waiting for binaries..."
-        break
-      fi
-      sleep 5
-    done
+  kube::log::error "Could not connect to rsync container. See build/README.md for setting up remote Docker engine."
+  return 1
+}
 
-    "${DOCKER[@]}" rm -f -v "${KUBE_BUILD_CONTAINER_NAME}" >/dev/null 2>&1 || true
-  else
-    kube::log::status "Output directory is local.  No need to copy results out."
+function kube::build::stop_rsyncd_container() {
+  V=6 kube::log::status "Stopping any currently running rsyncd container"
+  kube::build::destroy_container "${KUBE_RSYNC_CONTAINER_NAME}"
+}
+
+# This will launch rsyncd in a container and then sync the source tree to the
+# container over the local network.
+function kube::build::sync_to_container() {
+  kube::log::status "Syncing sources to container"
+
+  kube::build::start_rsyncd_container
+
+  local rsync_extra=""
+  if (( $KUBE_VERBOSE >= 6 )); then
+    rsync_extra="-iv"
+  fi
+
+  V=6 kube::log::status "Running rsync"
+  rsync $rsync_extra \
+    --archive \
+    --prune-empty-dirs \
+    --password-file="${LOCAL_OUTPUT_BUILD_CONTEXT}/rsyncd.password" \
+    --filter='- /.git/' \
+    --filter='- /.make/' \
+    --filter='- /_tmp/' \
+    --filter='- /_output/' \
+    --filter='- /' \
```

Comment:
This prevents it from explicitly trying to set uid/gid/perms on the root directory.  I'll add a comment.